### ISSUE TITLE
Crash on llvm::MemorySSAUpdater::applyInsertUpdates

### ISSUE THREAD

**Issue by BikingGlobetrotter** [llvm/llvm-project#54072]
Hi there,
I'd like to report a bug that occurs building our code base with the latest official clang release 13.0.1. 
Do know where the issues comes from?
```
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /usr/bin/clang++-13 -O3 -fPIC -fvisibility-inlines-hidden -fdata-sections -ffunction-sections -fexceptions -ftrapping-math -Wall -Wno-unknown-pragmas -Wno-logical-op-parentheses -Wno-inconsistent-missing-override -Wno-unused-value -Wno-braced-scalar-init -Wno-unknown-warning-option -Werror=macro-redefined -Wno-dtor-name -fvisibility=hidden -w -include../../../executor_0/checkouts/trunk/dev/libraries/aux_sources/public/mwTarget_Platform.hpp -include../../../executor_0/checkouts/trunk/dev/libraries/aux_sources/public/mwAvoid_Warnings.hpp -O2 -fopenmp=libomp -std=c++17 -DBOOST_NO_CXX11_UNICODE_LITERALS -DBOOST_STACKTRACE_BACKTRACE_INCLUDE_FILE=\"/usr/lib/gcc/x86_64-linux-gnu/9/include/backtrace.h\" -DBOOST_STACKTRACE_USE_BACKTRACE -DEIGEN_MPL2_ONLY -DMW_DISABLE_PRECOMPILED_HEADER -DUNICODE -D_UNICODE "-I/home/jenkins_slave/workspace/AutoTest - ClangBuild Trunk/cmake-clang/Binaries/generated" "-I/home/jenkins_slave/workspace/AutoTest - ClangBuild Trunk/cmake-clang/Binaries/generated/public" -I/home/jenkins_slave/executor_0/checkouts/trunk/dev/third_party/numbib/inc -I/home/jenkins_slave/executor_0/checkouts/trunk/dev/libraries/misc/include -I/home/jenkins_slave/executor_0/checkouts/trunk/dev/libraries/misc/public -I/home/jenkins_slave/executor_0/checkouts/trunk/dev/libraries/misc/../aux_sources/public -I/home/jenkins_slave/executor_0/checkouts/trunk/dev/libraries/mwCppLanguageSupport/public -I/home/jenkins_slave/executor_0/checkouts/trunk/dev/third_party/zlib/. -I/home/jenkins_slave/executor_0/checkouts/trunk/dev/libraries/mwstd/include -I/home/jenkins_slave/executor_0/checkouts/trunk/dev/libraries/mwstd/public -I/home/jenkins_slave/executor_0/checkouts/trunk/dev/third_party/tinyxml/. -isystem /home/jenkins_slave/executor_0/checkouts/trunk/dev/third_party/boost/. -DNDEBUG -c -o third_party/numbib/CMakeFiles/numbib.dir/lib/basis.cpp.o /home/jenkins_slave/executor_0/checkouts/trunk/dev/third_party/numbib/lib/basis.cpp
1.	<eof> parser at end of file
2.	Optimizer
 #0 0x00007f1a876d8791 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xd95791)
 #1 0x00007f1a876d6940 llvm::sys::RunSignalHandlers() (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xd93940)
 #2 0x00007f1a876d7e20 llvm::sys::CleanupOnSignal(unsigned long) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xd94e20)
 #3 0x00007f1a8761c035 (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xcd9035)
 #4 0x00007f1a8ffa03c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x153c0)
 #5 0x00007f1a888b47b7 llvm::MemorySSAUpdater::applyInsertUpdates(llvm::ArrayRef<llvm::cfg::Update<llvm::BasicBlock*> >, llvm::DominatorTree&, llvm::GraphDiff<llvm::BasicBlock*, false> const*) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1f717b7)
 #6 0x00007f1a888b20e7 llvm::MemorySSAUpdater::applyUpdates(llvm::ArrayRef<llvm::cfg::Update<llvm::BasicBlock*> >, llvm::DominatorTree&, bool) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1f6f0e7)
 #7 0x00007f1a880eaa2b llvm::LoopRotation(llvm::Loop*, llvm::LoopInfo*, llvm::TargetTransformInfo const*, llvm::AssumptionCache*, llvm::DominatorTree*, llvm::ScalarEvolution*, llvm::MemorySSAUpdater*, llvm::SimplifyQuery const&, bool, unsigned int, bool, bool) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x17a7a2b)
 #8 0x00007f1a883e1563 llvm::LoopRotatePass::run(llvm::Loop&, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>&, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1a9e563)
 #9 0x00007f1a89d72dcd (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x342fdcd)
#10 0x00007f1a883d3090 llvm::Optional<llvm::PreservedAnalyses> llvm::PassManager<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>::runSinglePass<llvm::Loop, std::unique_ptr<llvm::detail::PassConcept<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>, std::default_delete<llvm::detail::PassConcept<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&> > > >(llvm::Loop&, std::unique_ptr<llvm::detail::PassConcept<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>, std::default_delete<llvm::detail::PassConcept<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&> > >&, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>&, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&, llvm::PassInstrumentation&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1a90090)
#11 0x00007f1a883d2ea4 llvm::PassManager<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>::runWithoutLoopNestPasses(llvm::Loop&, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>&, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1a8fea4)
#12 0x00007f1a883d2872 llvm::PassManager<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>::run(llvm::Loop&, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>&, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1a8f872)
#13 0x00007f1a89d6659d (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x342359d)
#14 0x00007f1a883d3d31 llvm::FunctionToLoopPassAdaptor::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1a90d31)
#15 0x00007f1a89d7418d (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x343118d)
#16 0x00007f1a878404be llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function> >::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xefd4be)
#17 0x00007f1a89126efd (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x27e3efd)
#18 0x00007f1a88794256 llvm::CGSCCToFunctionPassAdaptor::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1e51256)
#19 0x00007f1a8912853d (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x27e553d)
#20 0x00007f1a88790cb8 llvm::PassManager<llvm::LazyCallGraph::SCC, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1e4dcb8)
#21 0x00007f1a8859edfd (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1c5bdfd)
#22 0x00007f1a88792ce6 llvm::DevirtSCCRepeatedPass::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1e4fce6)
#23 0x00007f1a8859f1dd (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1c5c1dd)
#24 0x00007f1a887922e7 llvm::ModuleToPostOrderCGSCCPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1e4f2e7)
#25 0x00007f1a8859f09d (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1c5c09d)
#26 0x00007f1a8783f1fe llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module> >::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xefc1fe)
#27 0x00007f1a8859c201 llvm::ModuleInlinerWrapperPass::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x1c59201)
#28 0x00007f1a89d7875d (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x343575d)
#29 0x00007f1a8783f1fe llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module> >::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xefc1fe)
#30 0x00007f1a8e1d466a (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x195c66a)
#31 0x00007f1a8e1ce6d4 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x19566d4)
#32 0x00007f1a8e492a60 (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x1c1aa60)
#33 0x00007f1a8d468854 clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0xbf0854)
#34 0x00007f1a8e48f251 clang::CodeGenAction::ExecuteAction() (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x1c17251)
#35 0x00007f1a8ecd3726 clang::FrontendAction::Execute() (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x245b726)
#36 0x00007f1a8ec4e106 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x23d6106)
#37 0x00007f1a8ed46596 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x24ce596)
#38 0x0000000000413189 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/bin/clang++-13+0x413189)
#39 0x00000000004111a6 (/usr/bin/clang++-13+0x4111a6)
#40 0x00007f1a8e94c012 (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x20d4012)
#41 0x00007f1a8761bd8d llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xcd8d8d)
#42 0x00007f1a8e94b8f0 clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x20d38f0)
#43 0x00007f1a8e91f2a0 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x20a72a0)
#44 0x00007f1a8e91f68a clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) const (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x20a768a)
#45 0x00007f1a8e93486e clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) (/usr/lib/x86_64-linux-gnu/libclang-cpp.so.13+0x20bc86e)
#46 0x00000000004109e3 main (/usr/bin/clang++-13+0x4109e3)
#47 0x00007f1a8642a0b3 __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/../csu/libc-start.c:342:3
#48 0x000000000040e23e _start (/usr/bin/clang++-13+0x40e23e)
clang: error: clang frontend command failed with exit code 139 (use -v to see invocation)
Ubuntu clang version 13.0.1-++20220120110924+75e33f71c2da-1~exp1~20220120231001.58
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
[code_and_script.zip](https://github.com/llvm/llvm-project/files/8139245/code_and_script.zip)

clang: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/basis-d252a9.cpp
clang: note: diagnostic msg: /tmp/basis-d252a9.sh
clang: note: diagnostic msg:

********************
```

**Comment by BikingGlobetrotter** [comment]
[code_and_script.zip](https://github.com/llvm/llvm-project/files/8140781/code_and_script.zip)

Sry. I forgot about that. I noticed that the issue seems to be fixed in clang-15.

**Comment by fhahn** [comment]
> Sry. I forgot about that. I noticed that the issue seems to be fixed in clang-15.

Thanks. So it sounds like the issue can be closed?

**Comment by lee218llnl** [comment]
FWIW, I am seeing a similar crash with clang 13.0.1. In my case, I found that clang 14.0.0 does not crash. Do we know if there was an explicit change/fix that explains this? I would hate to have the issue resurface again. I can provide my reproducer in this issue or in a separate one if someone wants to look at it.

**Comment by fhahn** [comment]
> FWIW, I am seeing a similar crash with clang 13.0.1. In my case, I found that clang 14.0.0 does not crash. Do we know if there was an explicit change/fix that explains this? I would hate to have the issue resurface again. I can provide my reproducer in this issue or in a separate one if someone wants to look at it.

There could be other changes that may hide the issue in end-to-end compilation via Clang. 

The best bet to make sure this is fixed is if you could extract an LLVM IR reproducer that crashes with `opt -loop-rotate`. Depending on the size of the input, you might be able to extract such a reproducer by adding `-mllvm  -print-before=loop-rotate -mllvm -print-module-scope`.

**Comment by lee218llnl** [comment]
Here is the output when adding those flags:

[out.txt](https://github.com/llvm/llvm-project/files/8643281/out.txt)

And here is the full source file (with the .txt extension added to the original file name)

[small.cpp.txt](https://github.com/llvm/llvm-project/files/8643285/small.cpp.txt)

I compiled with this line, although your include paths would need to be different:

`/usr/WS2/lee218/delete/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/llvm-13.0.1-apwanuzjr3dhzx2kauqfdrawgba7yj4a/bin/clang++ -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -disable-free -disable-llvm-verifier -discard-value-names -main-file-name small.cpp -ffp-exception-behavior=strict -internal-isystem /usr/tce/packages/oneapi/oneapi-2022.1/compiler/2022.0.1/linux/bin-llvm/../compiler/include -internal-isystem /usr/tce/packages/gcc/default/lib64/gcc/x86_64-unknown-linux-gnu/4.9.3/../../../../include/c++/4.9.3 -internal-isystem /usr/tce/packages/gcc/default/lib64/gcc/x86_64-unknown-linux-gnu/4.9.3/../../../../include/c++/4.9.3/x86_64-unknown-linux-gnu -internal-isystem /usr/tce/packages/gcc/default/lib64/gcc/x86_64-unknown-linux-gnu/4.9.3/../../../../include/c++/4.9.3/backward -internal-isystem /usr/tce/packages/oneapi/oneapi-2022.1/compiler/2022.0.1/linux/lib/clang/14.0.0/include -internal-isystem /usr/local/include -internal-isystem /usr/tce/packages/gcc/default/lib64/gcc/x86_64-unknown-linux-gnu/4.9.3/../../../../x86_64-unknown-linux-gnu/include -internal-externc-isystem /include -internal-externc-isystem /usr/include -O2 -o small.o -x c++ small.cpp -mllvm -print-before=loop-rotate -mllvm -print-module-scope >& out.txt
Segmentation fault (core dumped)`

Let me know if there is any additional information you would like me to add.

**Comment by fhahn** [comment]
@lee218llnl Thanks for the reproducer! The crash can be reproduce with `opt -passes='loop-mssa(loop-rotate)' ` and the provided out.txt: https://llvm.godbolt.org/z/h3aWq64vj

The issue has been fixed in 14.0 & current `main`.
